### PR TITLE
Compute capability fix

### PR DESCRIPTION
**Pull request by umar456**
Fixes checkAndSet

Description
-----------
Fixes an issue where the device compute capability is larger than
the supported maximum of the CUDA runtime used to build ArrayFire.
This happens for example when you run the Turing card with a CUDA
runtime of 9.0. The compute capability of Turing is 7.5 and the
maximum supported by the runtime is 7.0/7.2. Before this change
we were only checking the major compute capability and not checking
the minor version to set the max compute capability of the device.
This caused errors like:

```
In file src/backend/cuda/compile_module.cpp:266
NVRTC Error(5): NVRTC_ERROR_INVALID_OPTION
Log:
nvrtc: error: invalid value for --gpu-architecture (-arch)
```
This PR also updates the error messages for failure cases.

The utility header in cuda_fp16.hpp is not included automatically
in CUDA 9. Additionally we need to pass the
--device-as-default-execution-space flag to nvrtc for JIT and
non-JIT kernels

* The moduleKey is an size_t object so the maximum number of digits
  it can have is 20 so the format length for that value is updated
* The runtime check messages are always logged (but not displayed)
  Errors are still only thrown in debug modes
* Display the compute capability of the CUDA device along with
  its name and other stats

  example:
```
  Found device: Quadro T2000 (sm_75) (3.82 GB | ~3164.06 GFLOPs | 16 SMs)
```

Changes to Users
----------------
Better error messages and better support for newer devices with older CUDA toolkits

Checklist
---------

- [x] Rebased on latest master
- [x] Code compiles
- [x] Tests pass
- ~[ ] Functions added to unified API~
- ~[ ] Functions documented~

### CODE REVIEW

File: src/backend/cuda/compile_module.cpp [hunk 1]
```diff
@@ -143,20 +143,29 @@ Module compileModule(const string &moduleKey, const vector<string> &sources,
...
+            "cuda_fp16.h",
+        };
+        constexpr size_t numHeaders = extent<decltype(header_names)>::value;
+        array<const char *, numHeaders> headers = {
+            "",
             cuda_fp16_hpp,
             cuda_fp16_h,
         };
-        array<const char *, 2> header_names = {"cuda_fp16.hpp", "cuda_fp16.h"};
+        static_assert(headers.size() == numHeaders,
```
Comment (9prady9) [inline src/backend/cuda/compile_module.cpp:157]: :+1:

File: src/backend/cuda/compile_module.cpp [hunk 2]
```diff
@@ -230,11 +240,11 @@ Module compileModule(const string &moduleKey, const vector<string> &sources,
             sourceStrings[22].c_str(), sourceStrings[23].c_str(),
             sourceStrings[24].c_str(), sourceStrings[25].c_str(),
             sourceStrings[26].c_str(), sourceStrings[27].c_str(),
-        };
-        static_assert(extent<decltype(headers)>::value == NumHeaders,
+            sourceStrings[28].c_str()};
+        static_assert(extent<decltype(headers)>::value == numHeaders,
```
Comment (9prady9) [inline src/backend/cuda/compile_module.cpp:244]: :+1:

File: src/backend/cuda/compile_module.cpp [hunk 3]
```diff
@@ -246,6 +256,7 @@ Module compileModule(const string &moduleKey, const vector<string> &sources,
     vector<const char *> compiler_options = {
         arch.data(),
         "--std=c++14",
+        "--device-as-default-execution-space",
```
Comment (9prady9) [inline src/backend/cuda/compile_module.cpp:259]: Hm, I thought I added the unconditionally to nvrtc flags

File: src/backend/common/half.hpp [hunk 4]
```diff
@@ -832,9 +832,7 @@ class alignas(2) half {
 #endif
 
    public:
-#if CUDA_VERSION >= 10000
     AF_CONSTEXPR
```
Comment (9prady9) [inline src/backend/common/half.hpp:835]: I believe this was added to circumvent another issue.

File: src/backend/cuda/device_manager.cpp [hunk 5]
```diff
@@ -105,50 +126,66 @@ bool checkDeviceWithRuntime(int runtime, pair<int, int> compute) {
...
-            "cuda runtime compute version (%d.%d). Using %d.%d.";
-        fprintf(stderr, errMsg, originalCompute.first, originalCompute.second,
-                prop.first, prop.second, prop.first, prop.second);
-#endif
-    } else if (prop.first > Toolkit2MaxCompute[0].major) {
+        computeCapability.first >= tkitMaxCompute->major) {
+        int minorVersion = embeddedDevice ? tkitMaxCompute->embedded_minor
+                                          : tkitMaxCompute->minor;
+
+        if (computeCapability.second > minorVersion) {
```
Comment (9prady9) [inline src/backend/cuda/device_manager.cpp:146]: Does this condition actually help? I mean, if major version is greater, minor version doesn't matter I think.

Similar check in the next elseif give path

Comment (umar456) [reply src/backend/cuda/device_manager.cpp:146]: for 9.0 the max compute capability is 7.0 but my card is 7.5. This configuration causes compilation errors when building with nvrtc like the one I posted in my description.

File: src/backend/common/half.hpp (hunk 4 above)
Comment (umar456) [reply src/backend/common/half.hpp:835]: I think the problem was the --device-as-default-execution-space flag for jit kernels. It was treating this function as a host only function in that case causing errors.

File: src/backend/cuda/device_manager.cpp (hunk 5 above)
Comment (9prady9) [reply src/backend/cuda/device_manager.cpp:146]: Right, I didn't notice the equal sign in >=. My bad

**Comment by tvandera** [comment]
Hi,

I'm having a similar issue on this configuration:

```
ArrayFire v3.7.3 (CUDA, 64-bit Linux, build 59ac7b980)
Platform: CUDA Runtime 10.1, Driver: 460.32.03
[0] A100-SXM4-40GB, 40537 MB, CUDA Compute 8.0
```

This is the error:
```
In function cuda::Module common::compileModule(const string&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char
> >&, const std::vector<std::__cxx11::basic_string<char> >&, bool)
In file src/backend/cuda/compile_module.cpp:277
NVRTC Error(5): NVRTC_ERROR_INVALID_OPTION
Log:
nvrtc: error: invalid value for --gpu-architecture (-arch)
```

I do not understand what I should do to fix this.

**Comment by 9prady9** [comment]
@tvandera That is expected outcome given that v3.7.3 installers aren't built with CUDA 11 support. Please check the latest 3.8 release which has CUDA 11 support.

https://arrayfire.com/arrayfire-v3-8-release/

**Pull request merged**